File: src/main/java/dev/toastcie/mathtoastools/vector/intvector/Vector4Int.java
package dev.toastcie.mathtoastools.vector.intvector;

public class Vector4Int {
    //TODO auto edit xyz xyw xzw yzw (and xy xz xw yz yw zw) variables when x y z or w is edited

    public int x;
    public int y;
    public int z;
    public int w;

    public Vector4Int(int x, int y, int z, int w) {
        this.x = x;
        this.y = y;
        this.z = z;
        this.w = w;

    }

    public Vector4Int(Vector2Int xy, int z, int w) {
        this.x = xy.x;
        this.y = xy.y;
        this.z = z;
        this.w = w;

    }

    public Vector4Int(Vector3Int xyz, int w) {
        this.x = xyz.x;
        this.y = xyz.y;
        this.z = xyz.z;
        this.w = w;

    }

    public Vector4Int(Vector2Int xy, Vector2Int zw) {
        this.x = xy.x;
        this.y = xy.y;
        this.z = zw.x;
        this.w = zw.y;

    }

    //specials vectors

    public Vector2Int xy() {
        return new Vector2Int(this.x, this.y);
    }

    public Vector2Int xz() {
        return new Vector2Int(this.x, this.z);
    }

    public Vector2Int xw() {
        return new Vector2Int(this.x, this.w);
    }

    public Vector2Int yz() {
        return new Vector2Int(this.y, this.z);
    }

    public Vector2Int yw() {
        return new Vector2Int(this.y, this.w);
    }

    public Vector2Int zw() {
        return new Vector2Int(this.z, this.w);
    }

    public Vector3Int xyz() {
        return new Vector3Int(this.x, this.y, this.z);
    }

    public Vector3Int xyw() {
        return new Vector3Int(this.x, this.y, this.w);
    }

    public Vector3Int xzw() {
        return new Vector3Int(this.x, this.z, this.w);
    }

    public Vector3Int yzw() {
        return new Vector3Int(this.y, this.z, this.w);
    }

    //calculations methods

    public Vector4Int add(Vector4Int other) {
        return new Vector4Int(this.x + other.x, this.y + other.y, this.z + other.z, this.w + other.w);
    }

    public Vector4Int subtract(Vector4Int other) {
        return new Vector4Int(this.x - other.x, this.y - other.y, this.z - other.z, this.w - other.w);
    }

    public Vector4Int prod(int scalar) {
        return new Vector4Int(this.x * scalar, this.y * scalar, this.z * scalar, this.w * scalar);
    }

    public Vector4Int cross(Vector4Int other) {
        return new Vector4Int(this.y * other.z - this.z * other.y, this.z * other.x - this.x * other.z, this.x * other.y - this.y * other.x, 0);
    }

    public int dot(Vector4Int other) {
        return this.x * other.x + this.y * other.y + this.z * other.z + this.w * other.w;
    }
}

File: src/main/kotlin/dev/toastcie/mathtoastools/vector/intvector/Vector4Int.kt
package dev.toastcie.mathtoastools.vector.intvector

import kotlin.math.sqrt

/**
 * A 4D vector with integer values
 * @property x the x value of the vector
 * @property y the y value of the vector
 * @property z the z value of the vector
 * @property w the w value of the vector
 * @constructor creates a vector with the given values
 * @constructor creates a vector with the given values
 * @constructor creates a vector with the given values
 */
data class Vector4Int(var x:Int, var y:Int, var z:Int, var w:Int) {
    //constructors
    constructor(xy: Vector2Int, z: Int, w: Int) : this(xy.x, xy.y, z, w)

    constructor(xyz: Vector3Int, w: Int) : this(xyz.x, xyz.y, xyz.z, w)

    constructor(xy: Vector2Int, zw: Vector2Int) : this(xy.x, xy.y, zw.x, zw.y)

    //specials vectors
    /**
     * The x and y values of the vector
     */
    var xy: Vector2Int
        get() = Vector2Int(x, y)
        set(value) {
            x = value.x
            y = value.y
        }

    /**
     * The x and z values of the vector
     */
    var xz: Vector2Int
        get() = Vector2Int(x, z)
        set(value) {
            x = value.x
            z = value.y
        }

    /**
     * The x and w values of the vector
     */
    var xw: Vector2Int
        get() = Vector2Int(x, w)
        set(value) {
            x = value.x
            w = value.y
        }

    /**
     * The y and z values of the vector
     */
    var yz: Vector2Int
        get() = Vector2Int(y, z)
        set(value) {
            y = value.x
            z = value.y
        }

    /**
     * The y and w values of the vector
     */
    var yw: Vector2Int
        get() = Vector2Int(y, w)
        set(value) {
            y = value.x
            w = value.y
        }

    /**
     * The z and w values of the vector
     */
    var zw: Vector2Int
        get() = Vector2Int(z, w)
        set(value) {
            z = value.x
            w = value.y
        }

    /**
     * The x, y and z values of the vector
     */
    var xyz: Vector3Int
        get() = Vector3Int(x, y, z)
        set(value) {
            x = value.x
            y = value.y
            z = value.z
        }

    /**
     * The x, y and w values of the vector
     */
    var xyw: Vector3Int
        get() = Vector3Int(x, y, w)
        set(value) {
            x = value.x
            y = value.y
            w = value.z
        }

    /**
     * The x, z and w values of the vector
     */
    var xzw: Vector3Int
        get() = Vector3Int(x, z, w)
        set(value) {
            x = value.x
            z = value.y
            w = value.z
        }

    /**
     * The y, z and w values of the vector
     */
    var yzw: Vector3Int
        get() = Vector3Int(y, z, w)
        set(value) {
            y = value.x
            z = value.y
            w = value.z
        }

    //calculations methods
    /**
     * Adds two vectors
     * @param other the vector to add
     * @return the sum of the two vectors
     */
    operator fun plus(other: Vector4Int): Vector4Int = Vector4Int(x + other.x, y + other.y, z + other.z, w + other.w)
    /**
     * Adds a scalar to the vector
     * @param other the scalar to add
     * @return the sum of the vector and the scalar
     */
    operator fun plus(other: Int): Vector4Int = Vector4Int(x + other, y + other, z + other, w + other)

    /**
     * Subtracts two vectors
     * @param other the vector to subtract
     * @return the difference of the two vectors
     */
    operator fun minus(other: Vector4Int): Vector4Int = Vector4Int(x - other.x, y - other.y, z - other.z, w - other.w)
    /**
     * Subtracts a scalar to the vector
     * @param other the scalar to subtract
     * @return the difference of the vector and the scalar
     */
    operator fun minus(other: Int): Vector4Int = Vector4Int(x - other, y - other, z - other, w - other)

    /**
     * multiply a vectors by a scalar
     * @param other the scalar to multiply
     * @return the multiplication of the vector and the scalar
     */
    operator fun times(other: Int): Vector4Int = Vector4Int(x * other, y * other, z * other, w * other)

    /**
     * dot product of two vectors
     *
     * @param other the vector to dot
     * @return the dot product of the two vectors
     */
    infix fun dot(other: Vector4Int): Int = x * other.x + (y * other.y) + (z * other.z) + (w * other.w)
}